docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -0,0 +1,177 @@
+# ADR 047: Handling evidence from light client
+
+## Changelog
+* 18-02-2020: Initial draft
+* 24-02-2020: Second version
+
+## Context
+
+If the light client is under attack (either directly - lunatic / phantom
+validators or indirectly - fork on the main chain), it's supposed to halt and
+send evidence of misbehavior to the correct full node. Upon receiving an
+evidence, full node should punish malicious validators (if possible).
+
+## Decision
+
+When a light client sees two conflicting headers (`H1.Hash() != H2.Hash()`,
+`H1.Height == H2.Height`), both having 1/3+ of voting power of the currently
+trusted validator set, it will submit a `ConflictingHeadersEvidence` to all
+full nodes it's connected to.
+
+```go
+type ConflictingHeadersEvidence struct {
+  H1 types.SignedHeader
+  H2 types.SignedHeader
+}
+```
+
+When a full node receives the `ConflictingHeadersEvidence` evidence, it should
+a) validate it b) figure out if malicious behaviour is obvious (immediately
+slashable) or the fork accountability protocol needs to be started.
+
+### Validating headers
+
+Check both headers are valid (`ValidateBasic`), have
+the same height and signed by 1/3+ of known validator set.
+
+- Q: What if light client validator set is not equal to full node's validator
+  set (i.e. from full node's point of view both headers are not properly signed;
+  this includes the case where none of the two headers were committed on the
+  main chain)
+
+  Reject the evidence. It means light client is following a fork, but, hey, at
+  least it will halt.
+
+- Q: Don't we want to punish validators who signed something else even if they
+  have less or equal than 1/3?
+
+  No consensus so far. Ethan said no, Zarko said yes.
+  https://github.com/tendermint/spec/pull/71#discussion_r374210533
+
+### Figuring out if malicious behaviour is immediately slashable
+
+Let's say H1 was committed on the main chain. Intersect validator sets of H1
+and H2.
+
+* if there are signers(H2) that are not part of validators(H1), they misbehaved as
+they are signing protocol messages in heights they are not validators =>
+immediately slashable (#F4).
```

Comment:
Here also I don't understand. I looked back at the fork accountability spec a bit and it is not clear there to me either. What if there is a case of full-fork on chain where validator set diverged and some honest validators un-bonded on one branch. I think the phantom scenario is for the case where we can tell with certainty that H1 is from the canonical chain and H2 is from a soft-fork...but not clear to me how a full node can determine this by just looking at H1 and H2.

Reply:
> What if there is a case of full-fork on chain where validator set diverged and some honest validators un-bonded on one branch.

Then we'll punish them on our chain for voting on the forked chain.

Reply:
> validator set diverged

note this can only happen with 2/3+ of malicious validators